AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSDataOperationProxy.m

```diff
@@ -0,0 +1,246 @@
+              documentId:(NSString *)documentId
+            documentType:(Class)documentType
+                document:(id<MSSerializableDocument> _Nullable)document
+             baseOptions:(MSBaseOptions *_Nullable)baseOptions
+        cachedTokenBlock:(void (^)(MSCachedTokenCompletionHandler))cachedTokenBlock
+     remoteDocumentBlock:(void (^)(MSDocumentWrapperCompletionHandler))remoteDocumentBlock
+       completionHandler:(MSDocumentWrapperCompletionHandler)completionHandler {
+
+  // Get effective device time to live.
+  NSInteger deviceTimeToLive = baseOptions ? baseOptions.deviceTimeToLive : MSDataStoreTimeToLiveDefault;
+
+  // Validate current operation.
+  if (![MSDataOperationProxy isValidOperation:operation]) {
+    NSString *message = @"Operation is not supported";
+    MSLogError([MSDataStore logTag], message);
+    completionHandler([[MSDocumentWrapper alloc] initWithDataStoreErrorCode:MSACDataStoreLocalStoreError
+                                                               errorMessage:message
+                                                                 documentId:documentId]);
+    return;
+  }
+
+  // Retrieve a cached token.
+  cachedTokenBlock(^(MSTokensResponse *_Nullable tokens, NSError *_Nullable error) {
+    // Handle error.
+    if (error) {
+      // WIP: Append incoming error to message here. Also, should we issue a new code (like current code) or just forward incoming error.
+      NSString *message = @"Error while retrieving cached token, aborting operation";
+      MSLogError([MSDataStore logTag], message);
+      completionHandler([[MSDocumentWrapper alloc] initWithDataStoreErrorCode:MSACDataStoreLocalStoreError
+                                                                 errorMessage:message
+                                                                   documentId:documentId]);
+      return;
+    }
+
+    // Extract first token.
+    MSTokenResult *token = tokens.tokens[0];
+
+    // Retrieve a cached document.
+    MSDocumentWrapper *cachedDocument = [self.documentStore readWithToken:token documentId:documentId documentType:documentType];
+
+    // Execute remote operation if needed.
+    if ([self needsRemoteOperation:cachedDocument]) {
+      MSLogInfo([MSDataStore logTag], @"Performing remote operation");
+      remoteDocumentBlock(^(MSDocumentWrapper *_Nonnull remoteDocument) {
+        // If a valid remote document was retrieved, update local store
+        if (remoteDocument.error == nil) {
+          [self updateLocalStore:token
+              currentCachedDocument:cachedDocument
+                  newCachedDocument:remoteDocument
+                   deviceTimeToLive:deviceTimeToLive
+                          operation:operation];
+        }
+
+        // Complete the operation.
+        completionHandler(remoteDocument);
+      });
+    }
+
+    // Use cached document if possible.
+    else {
+
+      // Read operation.
+      if (operation == nil) {
```

Comment:
it would improve the readability if the caller passes something like `kMSReadOperation`